### PR TITLE
Gitlab

### PR DESCRIPTION
**Pull request by tuxnam**
Fixes #

Parser and analytics rules for GitLab Standalone edition.

## Proposed Changes
1 Solution for Gitlab imcluding 3 parsers for GitLab and 9 analytics rules for GitLab
 + 3 Custom tables for pipeline validation

### CODE REVIEW

**Comment by tuxnam** [comment]
Issues in the validation pipeline with npm?

**Comment by v-sabiraj** [comment]
Hey @tuxnam, We have resolved the issue for PR Validations failure and request you to please update your PR branch with the latest changes from **Azure:master** so that the PR Validations can be executed.... Thanks!!!

**Comment by tuxnam** [comment]
Hello,

it seems there are still issues in the validation.
I have custom functions but declared them in .scripts in KQLValidation folder. Seems to be another issue with KQL validation? Or did I miss something?

**Comment by v-sabiraj** [comment]
@aprakash13 please review hunting queries, thanks.

**Comment by v-jayakal** [comment]
@tuxnam - Rebased with master to run the pipeline to see the error as the earlier run was expired. Validations still show KQL test failure, will update you on that.

File: .script/tests/KqlvalidationsTests/CustomTables/GitLabAccess.json
Comment (v-jayakal) [inline .script/tests/KqlvalidationsTests/CustomTables/GitLabAccess.json:48]: end bracket '}' is missing, hence the validation error, please fix it.

**Review by v-jayakal — CHANGES_REQUESTED** [review]

File: .script/tests/KqlvalidationsTests/CustomTables/GitLabAccess.json
Comment (tuxnam) [reply .script/tests/KqlvalidationsTests/CustomTables/GitLabAccess.json:48]: done! good catch, thanks!

**Review by v-jayakal — APPROVED** [review]

**Comment by v-jayakal** [comment]
@tuxnam - I have approved after verification of Parsers. Please work with @aprakash13 or @shainw for analytics review. Thanks.

File: Hunting Queries/GitLab/GitLab_MaliciousIP
Comment (aprakash13) [inline Hunting Queries/GitLab/GitLab_MaliciousIP:26]: We probably need to check for the latest indicator as well as see if it is still active so that we do not provide wrong information by adding these to the query. 
  | summarize LatestIndicatorTime = arg_max(TimeGenerated, *) by IndicatorId
  | where Active == true

Please look and model the queries similar to this one: https://github.com/Azure/Azure-Sentinel/blob/master/Detections/ThreatIntelligenceIndicator/IPEntity_CustomSecurityLog.yaml

File: Hunting Queries/GitLab/GitLab_LocalAuthNoMFA
Comment (aprakash13) [inline Hunting Queries/GitLab/GitLab_LocalAuthNoMFA:21]: Can you please double check if the query runs correctly without errors and gives expected results?

**Review by aprakash13 — CHANGES_REQUESTED** [review]
Please confirm if these are detection or hunting queries. 

If these are **hunting queries** we do not need fields like 
queryFrequency: 1h
queryPeriod: 1d
triggerOperator: gt
triggerThreshold: 0
version: 1.0.0

If these are **detections**: 
Please change the query location to the detections folder.
Also add  'kind: Scheduled' property in the end.

For the sake of having a consistent format please also:
++ put two spaces in front of each line in the query body, to standardize on the ease of reading.
++ remove new line between end of query and statrt of entity mapping or version information

Have also added a few more suggestions. Please have a look. Thanks.

File: Hunting Queries/GitLab/GitLab_RepoVisibilityChange
Comment (tuxnam) [reply Hunting Queries/GitLab/GitLab_RepoVisibilityChange]: Adapted.

File: Hunting Queries/GitLab/GitLab_PAT_Repo
Comment (tuxnam) [reply Hunting Queries/GitLab/GitLab_PAT_Repo]: Good catch.

File: Hunting Queries/GitLab/GitLab_MaliciousIP
Comment (tuxnam) [reply Hunting Queries/GitLab/GitLab_MaliciousIP]: I already take latest indicators in the query, something I missed?

File: Hunting Queries/GitLab/GitLab_LocalAuthNoMFA
Comment (tuxnam) [reply Hunting Queries/GitLab/GitLab_LocalAuthNoMFA]: Yes double-checked.

**Comment by tuxnam** [comment]
Issues fixed and adapted as per comments.

**Comment by aprakash13** [comment]
@v-jayakal   Wondering, if this should be packaged as a solution. Thanks.

**Comment by v-sumdh** [comment]
@aprakash13 : can we package this as solution without the data connector. Please comment. Thanks

**Comment by tuxnam** [comment]
@v-sumdh @aprakash13 I am ok to do a data connector if needed to package it as a solution. It is syslog so not complex to do.

**Comment by aprakash13** [comment]
> @v-sumdh @aprakash13 I am ok to do a data connector if needed to package it as a solution. It is syslog so not complex to do.

Thanks @tuxnam / @v-sumdh and sorry for the delay in getting back on this. 
Considering Gitlab is such a popular management platform for Git repositories, it seemed like this would be better packaged as a solution, instead of being a bunch of detection/hunting queries. @tuxnam has agreed to help with the data connector and package this as a solution. @v-sumdh :  Can you please work with him to get this packaged as a solution. 

Here is the link to our guidance around creating solution that you might already have looked at.
https://github.com/Azure/Azure-Sentinel/tree/master/Solutions#step-1--create-your-content

**Comment by NikTripathi** [comment]
Hi @tuxnam Can you please move it to solution so we can go ahead with the solution packaging.

**Comment by NikTripathi** [comment]
Hi @aprakash13 Can you please have a look on detections and approve? Thanks

**Comment by NikTripathi** [comment]
@aprakash13 Can you please take a look and approve?

**Comment by tuxnam** [comment]
Hello,

Sorry I had huge workload recently and could not focus on this anymore.
I will build a data connector this week.
I will then package as a solution.

ETA: end of March.

feel free to propose other hunting or detection rules as well.

**Comment by tuxnam** [comment]
@v-sumdh @aprakash13 @v-jayakal  I did the first step to package as a solution. Can you validate as per documentation rules before I move to step 2 and ARM?

File: Solutions/GitLab/Analytic Rules/GitLab_ExternalUser.yaml
Comment (aprakash13) [inline Solutions/GitLab/Analytic Rules/GitLab_ExternalUser.yaml:16]: Please add relevantTechniques

File: Solutions/GitLab/Analytic Rules/GitLab_Impersonation.yaml
Comment (aprakash13) [inline Solutions/GitLab/Analytic Rules/GitLab_Impersonation.yaml:17]: Please add relevantTechniques

File: Solutions/GitLab/Analytic Rules/GitLab_RepoVisibilityChange.yaml
Comment (aprakash13) [inline Solutions/GitLab/Analytic Rules/GitLab_RepoVisibilityChange.yaml:17]: Please add relevantTechniques

File: Solutions/GitLab/Analytic Rules/GitLab_Repo_Deletion.yaml
Comment (aprakash13) [inline Solutions/GitLab/Analytic Rules/GitLab_Repo_Deletion.yaml:16]: Please add relevantTechniques

**Review by aprakash13 — CHANGES_REQUESTED** [review]
Overall, from a queries perspective it looks good. 
Please add relevant techniques to a few of the queries.
Once those are done feel free to move to the next stage if everything else looks good. Thanks.

**Comment by vmanojreddy** [comment]
@tuxnam Please move  all the files into solution folder. Thanks!!!

**Comment by vmanojreddy** [comment]
@tuxnam Please move  all the files into solution folder. Thanks!!!

**Review by tuxnam — COMMENTED** [review]
Reviewed techniques.
I created a solution folder. Not sure however about packaging, how I should build the ARM template? Any simple example?

**Comment by vmanojreddy** [comment]
@tuxnam  can you please maintain same names for both Parsers files and queries names for consistency

**Comment by tuxnam** [comment]
> @tuxnam can you please maintain same names for both Parsers files and queries names for consistency. Thanks!!!

@vmanojreddy : not sure what you mean here? thanks!

**Review by tuxnam — COMMENTED** [review]
Changes done for MITRE techniques not sure however what you meant by your request of name changes for consistency

**Comment by v-sabiraj** [comment]
> > @tuxnam can you please maintain same names for both Parsers files and queries names for consistency. Thanks!!!
> 
> @vmanojreddy : not sure what you mean here? thanks!

Hey @tuxnam , the parser will be saved as "GitLabAccess" as per instructions , but in the  data connector query, it is used as "GitlabAccess", where the letter "L" is in lowercase, same for GitlabAudit and GitlabApp, So either the query must be changed or the name of the parser should change, please check and update, thanks.
image

File: Solutions/GitLab/Analytic Rules/GitLab_BruteForce.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_BruteForce.yaml:32]: It would be best if you cast EventTime to datetime as part of the parser, which would mean the parser needs to parse the Message also.

File: Solutions/GitLab/Analytic Rules/GitLab_BruteForce.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_BruteForce.yaml:33]: Why rename this here, why not just name the field User as part of the parse above in line 29?

File: Solutions/GitLab/Analytic Rules/GitLab_ExternalUser.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_ExternalUser.yaml:27]: Why rename to UserAdded, when you can label the field in the parse from line 26?

File: Solutions/GitLab/Parsers/GitLabApp
Comment (shainw) [inline Solutions/GitLab/Parsers/GitLabApp:26]: Why not parse out Message field here instead of in each query?

File: Solutions/GitLab/Analytic Rules/GitLab_BruteForce.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_BruteForce.yaml:28]: I think I understand why you are using contains here versus has, but I think the best solution would be to use the parser you built to parse the Message field so that using contains here is not needed.

File: Solutions/GitLab/Analytic Rules/GitLab_ExternalUser.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_ExternalUser.yaml:25]: As stated before, having the parser do this?

File: Solutions/GitLab/Analytic Rules/GitLab_Impersonation.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_Impersonation.yaml:22]: Seems like the parser should be casting to string, then this doesn't need to be done in very query.

File: Solutions/GitLab/Analytic Rules/GitLab_PAT_Repo.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_PAT_Repo.yaml:33]: Parser should make EntityPath a string versus doing in query?

File: Solutions/GitLab/Analytic Rules/GitLab_PAT_Repo.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_PAT_Repo.yaml:24]: I don't think this is used in the query?

File: Solutions/GitLab/Analytic Rules/GitLab_PAT_Repo.yaml
Comment (shainw) [inline Solutions/GitLab/Analytic Rules/GitLab_PAT_Repo.yaml:27]: I don't think this is used in the query?

**Review by shainw — CHANGES_REQUESTED** [review]
Some items to look at related to moving query functions to parsers.

**Comment by vmanojreddy** [comment]
@shainw Please check and approve the changes, thanks.

**Comment by vmanojreddy** [comment]
@tuxnam In checks list Kqlvalidations failing due to the name 'FailedLogin' and EventTime does not refer to any known column, table, variable or function. thanks!!!

**Comment by tuxnam** [comment]
@vmanojreddy yes my bad, fixed. :)

**Comment by vmanojreddy** [comment]
@shainw Please check and approve the changes, thanks.

**Comment by vmanojreddy** [comment]
@shainw Please check and approve the changes, thanks.

**Comment by vmanojreddy** [comment]
@shainw Please check and approve the changes, thanks.

**Review by vmanojreddy — CHANGES_REQUESTED** [review]

**Review by aprakash13 — DISMISSED** [review]
Looks good from a query perspective. 
If there are any other feedbacks related to other components of this solution, please address them before merging.

**Comment by vmanojreddy** [comment]
@tuxnam Please the request the changes. thanks!!!

**Review by vmanojreddy — APPROVED** [review]

**Pull request merged**
